File: 2751. Robot Collisions.java
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        List<Robot> robots = new ArrayList<>();
        for (int i = 0; i < positions.length; i++) {
            robots.add(new Robot(i, positions[i], directions.charAt(i), healths[i]));
        }
        Collections.sort(robots, (a,b) -> a.position - b.position);
        Deque<Robot> stack = new ArrayDeque<>();
        loop: for (Robot r : robots) {
            if (stack.isEmpty() || r.direction == 'R') {
                stack.addLast(r);
                continue;
            }
            while (!stack.isEmpty() &&
                    stack.peekLast().direction == 'R') {
                if (stack.peekLast().health == r.health) {
                    stack.pollLast();
                    continue loop;
                } else if (stack.peekLast().health < r.health) {
                    stack.pollLast();
                    r.health -= 1;
                } else {
                    stack.peekLast().health -= 1;
                    continue loop;
                }
            }
            stack.addLast(r);
        }
        List<Robot> result = new ArrayList<>();
        while (!stack.isEmpty()) {
            result.add(stack.pollFirst());
        }
        Collections.sort(result, (a,b) -> a.index - b.index);
        List<Integer> answer = new ArrayList();
        for (Robot r : result) {
            answer.add(r.health);
        }
        return answer;
    }
}

class Robot {
    int index;
    int position;
    char direction;
    int health;
    Robot(int index, int position, char direction, int health) {
        this.index = index;
        this.position = position;
        this.direction = direction;
        this.health = health;
    }
}

File: 2751. Robot Collisions.go
func survivedRobotsHealths(positions []int, healths []int, directions string) []int {
	robots := make([]*robot, len(positions))
	for i := 0; i < len(positions); i++ {
		robots[i] = newRobot(i, positions[i], healths[i], directions[i])
	}
	sort.Slice(robots, func(i, j int) bool {
		return robots[i].position < robots[j].position
	})

	stack := make(stack, 0)

loop:
	for _, r := range robots {
		if len(stack) == 0 || r.direction == 'R' {
			stack.push(r)
			continue
		}
		for len(stack) > 0 && stack.peek().direction == 'R' {
			if stack.peek().health == r.health {
				stack.pop()
				continue loop
			}
			if stack.peek().health < r.health {
				stack.pop()
				r.health -= 1
			} else {
				stack.peek().health -= 1
				continue loop
			}
		}
		stack.push(r)
	}
	result := []*robot{}
	for len(stack) > 0 {
		result = append(result, stack.pop())
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].index < result[j].index
	})
	answer := []int{}
	for _, r := range result {
		answer = append(answer, r.health)
	}
	return answer
}

type robot struct {
	index     int
	position  int
	direction byte
	health    int
}

func newRobot(index, position, health int, direction byte) *robot {
	return &robot{index, position, direction, health}
}

type stack []*robot

func (s *stack) push(r *robot) {
	*s = append(*s, r)
}

func (s *stack) pop() *robot {
	if len(*s) == 0 {
		return nil
	}
	r := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return r
}

func (s *stack) peek() *robot {
	if len(*s) == 0 {
		return nil
	}
	return (*s)[len(*s)-1]
}